subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -0,0 +1,2482 @@
+		zoap_header_set_token(out.out_zpkt, token, tkl);
+	}
+
+	in.inpos = 0;
+	in.inbuf = zoap_packet_get_payload(in.in_zpkt, &in.insize);
+
+	/* TODO: check for block transfer? */
+
+	switch (context.operation) {
+
+	case LWM2M_OP_READ:
+		if (observe == 0) {
+			/* add new observer */
+			if (token) {
+				r = zoap_add_option_int(out.out_zpkt,
+							ZOAP_OPTION_OBSERVE, 1);
+				if (r) {
+					SYS_LOG_ERR("OBSERVE option error: %d",
+						    r);
+				}
+
+				r = engine_add_observer(
+					net_pkt_context(in.in_zpkt->pkt),
+					from_addr, token, tkl, &path);
+				if (r < 0) {
+					SYS_LOG_ERR("add OBSERVE error: %d", r);
+				}
+			} else {
+				SYS_LOG_ERR("OBSERVE request missing token");
+			}
+		} else if (observe == 1) {
+			/* use token from this request */
+			token = zoap_header_get_token(in.in_zpkt, &tkl);
+			/* remove observer */
+			r = engine_remove_observer(token, tkl);
+			if (r < 0) {
+				SYS_LOG_ERR("remove obserer error: %d", r);
+			}
+		}
+
+		/* set output content-format */
+		r = zoap_add_option_int(out.out_zpkt,
+					ZOAP_OPTION_CONTENT_FORMAT, accept);
+		if (r > 0) {
+			SYS_LOG_ERR("Error setting response content-format: %d",
+				    r);
+		}
+
+		r = do_read_op(obj, &context);
+		break;
+
+	case LWM2M_OP_DISCOVER:
+		r = do_discover_op(&context);
+		break;
+
+	case LWM2M_OP_WRITE:
+		r = do_write_op(obj, &context, format);
+		break;
+
+	case LWM2M_OP_WRITE_ATTR:
+		r = engine_write_attr_handler(obj, &context);
+		break;
+
+	case LWM2M_OP_EXECUTE:
+		r = engine_exec_handler(obj, &context);
+		break;
+
+	case LWM2M_OP_DELETE:
+		r = engine_delete_handler(obj, &context);
+		break;
+
+	default:
+		SYS_LOG_ERR("Unknown operation: %u", context.operation);
+		return -EINVAL;
+	}
+
+	if (r == 0) {
+		/* TODO: Handle blockwise 1 */
+
+		if (out.outlen > 0) {
+			SYS_LOG_DBG("replying with %u bytes", out.outlen);
+			zoap_packet_set_used(out.out_zpkt, out.outlen);
+		} else {
+			SYS_LOG_DBG("no data in reply");
+		}
+	} else {
+		if (r == -ENOENT) {
+			zoap_header_set_code(out.out_zpkt,
+					     ZOAP_RESPONSE_CODE_NOT_FOUND);
+			r = 0;
+		} else if (r == -EPERM) {
+			zoap_header_set_code(out.out_zpkt,
+					     ZOAP_RESPONSE_CODE_NOT_ALLOWED);
+			r = 0;
+		} else {
+			/* Failed to handle the request */
+			zoap_header_set_code(out.out_zpkt,
+					     ZOAP_RESPONSE_CODE_INTERNAL_ERROR);
+			r = 0;
+		}
+	}
+
+	return r;
+}
+
+static void udp_receive(struct net_context *ctx, struct net_pkt *pkt,
+			int status, void *user_data)
+{
+	struct zoap_pending *pending;
+	struct zoap_reply *reply;
+	struct zoap_packet response;
+	struct sockaddr from_addr;
+	struct zoap_packet response2;
+	struct net_pkt *pkt2;
+	int header_len, r;
+	const u8_t *token;
+	u8_t tkl;
+
+	/* Save the from address */
+#if defined(CONFIG_NET_IPV6)
+	if (net_pkt_family(pkt) == AF_INET6) {
+		net_ipaddr_copy(&net_sin6(&from_addr)->sin6_addr,
+				&NET_IPV6_HDR(pkt)->src);
+		net_sin6(&from_addr)->sin6_port = NET_TCP_HDR(pkt)->src_port;
```

Comment:
NET_TCP_HDR() -> NET_UDP_HDR()
Assuming we are using UDP.

And in latest master branch, both have been removed check 3604c391e69d253fb985aa6d072283250b454e6e